crates/swc_ecma_transforms_react/src/refresh/hook.rs

```diff
@@ -0,0 +1,523 @@
+use std::mem;
+
+use indexmap::IndexSet;
+use once_cell::sync::Lazy;
+use regex::Regex;
+use sha1::{Digest, Sha1};
+use swc_atoms::JsWord;
+use swc_common::{util::take::Take, SourceMap, Spanned, DUMMY_SP};
+use swc_ecma_ast::*;
+use swc_ecma_utils::{private_ident, quote_ident};
+use swc_ecma_visit::{
+    noop_visit_mut_type, noop_visit_type, Visit, VisitMut, VisitMutWith, VisitWith,
+};
+
+use crate::RefreshOptions;
+
+use super::util::{is_builtin_hook, make_call_expr, make_call_stmt, CollectIdent};
+
+// function that use hooks
+struct HookSig {
+    handle: Ident,
+    // need to add an extra register, or alreay inlined
+    hooks: Vec<Hook>,
+}
+
+impl HookSig {
+    fn new(hooks: Vec<Hook>) -> Self {
+        HookSig {
+            handle: private_ident!("_s"),
+            hooks,
+        }
+    }
+}
+
+struct Hook {
+    callee: HookCall,
+    key: String,
+}
+
+// we only consider two kinds of callee as hook call
+enum HookCall {
+    Ident(Ident),
+    Member(Expr, Ident), // for obj and prop
+}
+pub struct HookRegister<'a> {
+    pub options: &'a RefreshOptions,
+    pub ident: Vec<Ident>,
+    pub extra_stmt: Vec<Stmt>,
+    pub scope_binding: IndexSet<JsWord>,
+    pub cm: &'a SourceMap,
+    pub should_reset: bool,
+}
+
+impl<'a> HookRegister<'a> {
+    pub fn gen_hook_handle(&mut self) -> Stmt {
+        Stmt::Decl(Decl::Var(VarDecl {
+            span: DUMMY_SP,
+            kind: VarDeclKind::Var,
+            decls: self
+                .ident
+                .take()
+                .into_iter()
+                .map(|id| VarDeclarator {
+                    span: DUMMY_SP,
+                    name: Pat::Ident(BindingIdent::from(id)),
+                    init: Some(Box::new(make_call_expr(quote_ident!(self
+                        .options
+                        .refresh_sig
+                        .clone())))),
+                    definite: false,
+                })
+                .collect(),
+            declare: false,
+        }))
+    }
+
+    // The second call is around the function itself. This is used to associate a
+    // type with a signature.
+    // Unlike with $RefreshReg$, this needs to work for nested declarations too.
+    fn wrap_with_register(&self, handle: Ident, func: Expr, hooks: Vec<Hook>) -> Expr {
+        let mut args = vec![func];
```

Comment:
I think changing the type of `args` to `Vec<ExprOrSpread>` will be better for performance as it reduces memory operations.

Reply:
You can use `let mut args = vec![func.as_arg()];` instead.